engine/common/rpc/convert/convert.go

```diff
@@ -814,3 +816,159 @@ func MessageToChunk(m *entities.Chunk) (*flow.Chunk, error) {
+	collection := &entities.ExecutionDataCollection{}
+	if data.Collection != nil {
+		collection = &entities.ExecutionDataCollection{
+			Transactions: TransactionsToMessages(data.Collection.Transactions),
+		}
+	}
+
+	events := EventsToMessages(data.Events)
+	if len(events) == 0 {
+		events = nil
+	}
+
+	paths := make([][]byte, len(data.TrieUpdate.Paths))
+	for i, path := range data.TrieUpdate.Paths {
+		paths[i] = path[:]
+	}
+
+	payloads := make([]*entities.Payload, len(data.TrieUpdate.Payloads))
+	for i, payload := range data.TrieUpdate.Payloads {
+		key, err := payload.Key()
+		if err != nil {
+			return nil, err
+		}
+		keyParts := make([]*entities.KeyPart, len(key.KeyParts))
+		for j, keyPart := range key.KeyParts {
+			keyParts[j] = &entities.KeyPart{
+				Type:  uint32(keyPart.Type),
+				Value: keyPart.Value,
+			}
+		}
+		payloads[i] = &entities.Payload{
+			KeyPart: keyParts,
+			Value:   payload.Value(),
+		}
+	}
+
+	trieUpdate := &entities.TrieUpdate{
+		RootHash: data.TrieUpdate.RootHash[:],
+		Paths:    paths,
+		Payloads: payloads,
+	}
+
+	return &entities.ChunkExecutionData{
+		Collection: collection,
+		Events:     events,
+		TrieUpdate: trieUpdate,
+	}, nil
+}
+
+func MessageToBlockExecutionData(m *entities.BlockExecutionData, chain flow.Chain) (execution_data.BlockExecutionData, error) {
+	if m == nil {
+		return execution_data.BlockExecutionData{}, ErrEmptyMessage
+	}
+	chunks := make([]*execution_data.ChunkExecutionData, len(m.ChunkExecutionData))
+	for i, chunk := range m.GetChunkExecutionData() {
+		convertedChunk, err := MessageToChunkExecutionData(chunk, chain)
+		if err != nil {
+			return execution_data.BlockExecutionData{}, err
+		}
+		chunks[i] = &convertedChunk
+	}
+
+	return execution_data.BlockExecutionData{
+		BlockID:             MessageToIdentifier(m.GetBlockId()),
+		ChunkExecutionDatas: chunks,
+	}, nil
+}
+
+func MessageToChunkExecutionData(m *entities.ChunkExecutionData, chain flow.Chain) (execution_data.ChunkExecutionData, error) {
+	collection, err := messageToExecutionDataCollection(m.GetCollection(), chain)
+	if err != nil {
+		return execution_data.ChunkExecutionData{}, err
+	}
+
+	trieUpdate, err := messageToTrieUpdate(m.GetTrieUpdate())
+	if err != nil {
+		return execution_data.ChunkExecutionData{}, err
+	}
+
+	events := MessagesToEvents(m.GetEvents())
+	if len(events) == 0 {
+		events = nil
+	}
+
+	return execution_data.ChunkExecutionData{
+		Collection: collection,
+		Events:     events,
+		TrieUpdate: trieUpdate,
+	}, nil
+}
+
+func messageToExecutionDataCollection(m *entities.ExecutionDataCollection, chain flow.Chain) (*flow.Collection, error) {
+	messages := m.GetTransactions()
+	transactions := make([]*flow.TransactionBody, len(messages))
+	for i, message := range messages {
+		transaction, err := MessageToTransaction(message, chain)
+		if err != nil {
+			return &flow.Collection{}, err
```

Comment:
any reason to return an actual object?
```suggestion
			return nil, err
```